bigtable_input.py

```diff
@@ -45,22 +47,66 @@ def get_game_range_row_names(game_begin, game_end):
     Sample row name:
       g_0000000001_m001
 
-    To capture the range of all moves in the two given games,
-    the end row will need to go up to g_00..(N+1).
-
     Args:
       game_begin:  an integer of the beginning game number.
-      game_end:  an integer of the ending game number, inclusive.
+      game_end:  an integer of the ending game number, exclusive.
 
     Returns:
       The two string row numbers to pass to Bigtable as the row range.
     """
     row_fmt = 'g_{:0>10}_'
-    return row_fmt.format(game_begin), row_fmt.format(game_end + 1)
+    return row_fmt.format(game_begin), row_fmt.format(game_end)
+
+
+def make_single_array(ds, batch_size=8*1024):
+    full_result = None
+    with tf.Session() as sess:
+        ds = ds.batch(batch_size)
+        iterator = ds.make_initializable_iterator()
+        sess.run(iterator.initializer)
+        get_next = iterator.get_next()
+        try:
+            while True:
+                result = sess.run(get_next)
+                if full_result is None:
+                    full_result = result
+                else:
+                    full_result = np.concatenate((full_result, result))
```

Comment:
in my experience this can be slow when you are constantly concatenating
maybe preallocate a big number then shrink the np.array afterwards?
I couldn't find a better link than this https://stackoverflow.com/a/13220868

Reply:
Collecting a list of np.array and then concatenating them is a factor of 2 overhead. Agreed that this method is O(n^2).

Reply:
I like the idea in the link you shared, of collecting a list of smaller arrays and then concatenating them in one call at the end. I'll do that.